clang/lib/AST/ExprConstant.cpp

```diff
@@ -15154,6 +15154,7 @@ static ICEDiag CheckICE(const Expr* E, const ASTContext &Ctx) {
   case Expr::CoawaitExprClass:
   case Expr::DependentCoawaitExprClass:
   case Expr::CoyieldExprClass:
+  case Expr::UniqueStableNameExprClass:
```

Comment:
This change is correct, but we should have some `constexpr` tests to make sure this works in constant expression evaluations. e.g., it would probably be useful to be able to write `static_assert(std::string(__builtin_unique_stable_name(whatever)) == "foo");`

Reply:
So that static assert actually wouldn't work for 2 reasons: 1- its a std::string, which isn't constexpr yet :)  But more importantly, #2, the value of __builtin_unique_stable_name isn't calculable until we are 'after' constexpr evaluation.  Basically, we have to see every template instantiation (and thus the entire TU evaluated) before we can actually 'know' the name.  Otherwise lambda numbers are going to be 'off'.

The previous version didn't need the knowledge of the entire TU (that is, of every lambda required to name a kernel) to calculate it, so it worked as a constant expression.

Reply:
https://en.cppreference.com/w/cpp/string/basic_string/basic_string sure make it seem like it is constexpr since C++20 (see signature #5)

However, the point about having to do this after constant evaluation time is interesting. I'll have to think on that a bit, but IIRC, there were some comments in the review that made me think constexpr was expected to work (so we may need a pass to make sure those comments get corrected).

Reply:
Oooh, neat! Didn't realize they'd gotten to string for 20, much industrious LEWG/LWG!

Not sure what comments you mean, but we'll have to discuss it at that point I think.  Thinking further, I hope the lack of constexpr doesn't break some of our use cases.

Reply:
@romanovvlad : Can you comment on this?  Do we need constexpr evaluation for this builtin?

Reply:
@erichkeane
Do not see why it would be necessary from the functional point of view.
On the other hand this can change the pattern which SpecConstants pass should recognize.
@vmaksimo  Could you please tell if it's OK for the pass?

Reply:
> Do not see why it would be necessary from the functional point of view.

I also can't think of a point where at DPC++ level `constexpr` matters here

> On the other hand this can change the pattern which SpecConstants pass should recognize.

I'm not exactly sure here - I assume that it would be the same string literal as it is now, which will be located in a global variable and references through some load/stores/geps/addrspacecasts.
It is hard to say without an LLVM IR example

Reply:
The constexprness means you couldn't use the builtin in a constexpr function, and you couldn't use the result of it as a template.  We were just making sure that is OK.

Reply:
e.g., the `static_assert` example I gave above would fail to compile (rather than perform the check and then potentially fail). That was about the only useful case I could think of, and it seems reasonable to not support it until someone needs it.

Reply:
@schittir -- the `static_assert` example would be a good one to add to the new test cases as well.